### DISCUSSION TITLE
Install dbatools on a computer without internet access

### DISCUSSION THREAD

**Discussion by andreasjordan** [dataplat/dbatools#7984] [Show and tell]
If you need to install dbatools on a computer without internet access, you only need two files: The NuGet-DLL and the dbatools-NuGet-package. 

You can get the NuGet-DLL Microsoft.PackageManagement.NuGetProvider.dll from any other system where you installed NuGet with an internet connection or download it from: https://onegetcdn.azureedge.net/providers/Microsoft.PackageManagement.NuGetProvider-2.8.5.208.dll and rename the file.

You download the dbatools-NuGet-package from https://www.powershellgallery.com/packages/dbatools using the "Manual Download":
![image](https://user-images.githubusercontent.com/66946165/143559008-5153c4e0-a74b-4c32-85e1-568fd5c79e55.png)

Then you need to transfer those two files to the target computer and only have to change the following code to your needs:

```powershell
$nugetPath = "$($env:ProgramFiles)\PackageManagement\ProviderAssemblies\NuGet\2.8.5.208"
$null = New-Item -Path $nugetPath -Type Directory
Copy-Item -Path .\Microsoft.PackageManagement.NuGetProvider.dll -Destination $nugetPath

$repoPath = "C:\Temp\MyPSGallery"
$null = New-Item -Path $repoPath -Type Directory
Copy-Item -Path .\dbatools.1.1.38.nupkg -Destination $repoPath
Register-PSRepository -Name MyPSGallery -SourceLocation $repoPath -PublishLocation $repoPath -InstallationPolicy Trusted

Install-Module -Name dbatools -Repository MyPSGallery
Import-Module -Name dbatools
Get-Module -Name dbatools
```

Let me share my screenshot as a proof of concept:
![OfflineInstall](https://user-images.githubusercontent.com/66946165/143559403-b0a2b3e3-a2a4-4d1a-b76c-a31932c34bd4.JPG)

I worked in a remote session on a windows server core, so to run it locally you probably need an elevated PowerShell.

See also my blog post for even more information: https://blog.ordix.de/installation-and-use-of-dbatools-on-a-computer-without-internet-connection

**Comment by nanoDBA** [comment]
Thank you for taking the time to write this up and share it here from your original excellent tweet.  Much appreciated!

**Comment by andreasjordan** [comment]
You want to update dbatools to a new version? No problem. Just download the new nupkg file and place it next to the old in you local repository. Then you can use `Update-Module -Name dbatools` to install the new version.

**Comment by potatoqualitee** [comment]
Awesome, thank you so much andreas!

**Comment by andreasjordan** [comment]
See also my blog post about that: https://blog.ordix.de/installation-and-use-of-dbatools-on-a-computer-without-internet-connection

**Comment by bneal1987** [comment]
Walking through this I am getting an error and was hoping you could assist with this. 

<img width="997" alt="Screenshot 2023-05-04 at 11 37 01 AM" src="https://user-images.githubusercontent.com/82835140/236257810-a83a94ef-9179-4843-b999-0b72a1611bdb.png">

Thanks,

**Reply by bneal1987** [reply]
Any ideas on the above error?

Thanks,

**Reply by andreasjordan** [reply]
Sorry, no. The first error tells you that the directory is already there - so NuGet might already be set up correctly. You probably should not have copied the ddl there. The second error tells you that the command Register-PackageSource is different from what we expect, as the property SourceLocation is not found. Looks like your environment is different than mine - so the code will not help you.

**Reply by bneal1987** [reply]
Thanks you for the quick response on this!
